Mathlib/Analysis/CStarAlgebra/Module/Constructions.lean

```diff
@@ -0,0 +1,414 @@
+      exact le_add_of_nonneg_right CStarModule.inner_self_nonneg |>.trans_eq h
+    · refine inner_self.mp <| le_antisymm ?_ inner_self_nonneg
+      exact le_add_of_nonneg_left CStarModule.inner_self_nonneg |>.trans_eq h
+  inner_op_smul_right := by simp [add_mul]
+  inner_smul_right_complex := by simp [smul_add]
+  star_inner x y := by simp
+  norm_eq_sqrt_norm_inner_self {x} := by with_reducible_and_instances rfl
+
+lemma prod_inner (x y : C⋆ᵐᵒᵈ (E × F)) : ⟪x, y⟫_A = ⟪x.1, y.1⟫_A + ⟪x.2, y.2⟫_A := rfl
+
+variable [CStarRing A] [SMulCommClass ℂ A A] [IsScalarTower ℂ A A] [CompleteSpace A]
+
+lemma max_le_prod_norm (x : C⋆ᵐᵒᵈ (E × F)) : max ‖x.1‖ ‖x.2‖ ≤ ‖x‖ := by
+  rw [prod_norm]
+  simp only [equiv_fst, norm_eq_sqrt_norm_inner_self, equiv_snd, max_le_iff, norm_nonneg,
+    Real.sqrt_le_sqrt_iff]
+  constructor
+  all_goals
+    apply norm_le_norm_of_nonneg_of_le
+    all_goals
+      aesop (add safe apply CStarModule.inner_self_nonneg)
+
+lemma norm_equiv_le_norm_prod (x : C⋆ᵐᵒᵈ (E × F)) : ‖equiv (E × F) x‖ ≤ ‖x‖ :=
+  max_le_prod_norm x
+
+section Aux
+
+attribute [-instance] WithCStarModule.instUniformSpace WithCStarModule.instBornology
+attribute [local instance] CStarModule.normedAddCommGroup
+
+open Filter Uniformity Bornology
+
+lemma antilipschitzWith_two_equiv_prod_aux : AntilipschitzWith 2 (equiv (E × F)) :=
+  AddMonoidHomClass.antilipschitz_of_bound (linearEquiv ℂ (E × F)) fun x ↦ by
+    apply prod_norm_le_norm_add x |>.trans
+    simp only [NNReal.coe_ofNat, linearEquiv_apply, two_mul]
+    gcongr
+    · exact norm_fst_le x
+    · exact norm_snd_le x
+
+lemma lipschitzWith_one_equiv_prod_aux : LipschitzWith 1 (equiv (E × F)) :=
+  AddMonoidHomClass.lipschitz_of_bound_nnnorm (linearEquiv ℂ (E × F)) 1 <| by
+    simpa using norm_equiv_le_norm_prod
+
+lemma uniformity_prod_eq_aux :
+    𝓤[(inferInstance : UniformSpace (E × F)).comap <| equiv _] = 𝓤 (C⋆ᵐᵒᵈ (E × F)) :=
+  uniformity_eq_of_bilipschitz antilipschitzWith_two_equiv_prod_aux lipschitzWith_one_equiv_prod_aux
+
+lemma isBounded_prod_iff_aux (s : Set (C⋆ᵐᵒᵈ (E × F))) :
+    @IsBounded _ (induced <| equiv (E × F)) s ↔ IsBounded s :=
+  isBounded_iff_of_bilipschitz antilipschitzWith_two_equiv_prod_aux
+    lipschitzWith_one_equiv_prod_aux s
+
+end Aux
+
+noncomputable instance : NormedAddCommGroup (C⋆ᵐᵒᵈ (E × F)) :=
+  .ofCoreReplaceAll normedSpaceCore uniformity_prod_eq_aux isBounded_prod_iff_aux
+
+instance : NormedSpace ℂ (C⋆ᵐᵒᵈ (E × F)) := .ofCore normedSpaceCore
+
+end Prod
+
+/-! ## Pi-types of C⋆-modules -/
+
+section Pi
+
+variable {ι : Type*} {E : ι → Type*} [Fintype ι]
+variable [∀ i, NormedAddCommGroup (E i)] [∀ i, Module ℂ (E i)] [∀ i, SMul Aᵐᵒᵖ (E i)]
+variable [∀ i, CStarModule A (E i)]
+
+noncomputable instance : Norm (C⋆ᵐᵒᵈ (Π i, E i)) where
+  norm x := √‖∑ i, ⟪x i, x i⟫_A‖
+
+lemma pi_norm (x : C⋆ᵐᵒᵈ (Π i, E i)) : ‖x‖ = √‖∑ i, ⟪x i, x i⟫_A‖ := by
+  with_reducible_and_instances rfl -- this would fail without `⇑`, ensures no defeq abuse
```

Comment:
I don't understand the comment here (or the need for `with_reducible_and_instances`, it does work without it).

Reply:
oops, yes, also a result of a previous version before I had the Zulip discussion about defeq abuse.